File: 0031-next-permutation/0031-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int brp=nums.size()-2;
        while(brp >= 0) {
            if(nums[brp] < nums[brp+1])
                break;
            brp--;
        }
        int loc=nums.size()-1;
        if(brp != -1){
            while(loc >= brp){
                if(nums[loc] > nums[brp])
                    break;
                loc--;
            }
            swap(nums[brp],nums[loc]);
        }
        reverse(nums.begin()+brp+1,nums.end());
    }
};

File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    public void swap(int []arr,int i,int j){
        int temp = arr[i];
        arr[i]=arr[j];
        arr[j]=temp;
    }
    
    public void nextPermutation(int[] nums) {
        int n = nums.length;
        int idx = -1;
        for(int i=n-2;i>=0;i--){
            if(nums[i] < nums[i+1]){
                idx=i;
                break;
            }
        }
        if(idx == -1){
            Arrays.sort(nums);
            return;
        }
        int greaterNum = 0;
        for(int i=n-1;i>idx;i--){
            if(nums[i] > nums[idx]){
                greaterNum = i;
                break;
            }
        }
        swap(nums,idx,greaterNum);
        Arrays.sort(nums,idx+1,n);
    }
}